### ISSUE TITLE
Display current consumption and power to the grid

### ISSUE THREAD

**Issue by eSportler89** [Kubiac/homebridge-solaredge-realtime#2]
**Is your feature request related to a problem? Please describe:**
It's just a feature request to get more possibilities to trigger an event.

**Describe the solution you'd like:**
Showing current consumption and power to the grid the same way as produced power.

**Describe alternatives you've considered:**
--

**Additional context:**
I want to trigger my go-e Charger Wallbox when there is a specific amount of power going to the grid.
Tanks :)

**Comment by Kubiac** [comment]
Hi, so I assume you have a solaredge energy meter connected to your inverter? And you want to get data from that, correct?
I don't have an solaredge energy meter, so you would need to test this feature when implemented.

**Comment by eSportler89** [comment]
Yes, that's correct.
Testing is no problem :)

**Comment by eSportler89** [comment]
For now, no obvious errors are displayed, but I still only have one lamp as a HomeKit item. I would have expected three more lamps now.

**Comment by Kubiac** [comment]
Klar, gerne Deutsch.
Also, dass nur ein Lichtsensor angezeigt wird, ist erwartet. Wenn du an Produktion und Export/import interessiert bist, dann einfach 2 inverter mit jeweils meter = 0 und meter =1 konfigurieren.
wenn er aber immer 0,00001 anzeigt stimmt was nicht mit der Berechnung. Ich guck es mir nochmal an und melde mich. Im Homebridge-Log ist nichts zu sehen sagst du ja.

**Comment by eSportler89** [comment]
Okay, da hätte ich selbst drauf kommen können :D
Korrekt, im Logging ist leider gar nichts zu erkennen.
Ich weiß aber das es zu Beginn mal funktioniert hatte und ich zu jeder Änderung einen Logeintrag sehen konnte, nur kann ich dir nicht sagen ob ich danach mal ein Update deines Plugins gemacht hatte.

**Comment by Morcman** [comment]
Klar, ich unterstütze gerne. Meld dich einfach wenn du soweit bist.

**Comment by Kubiac** [comment]
Ok, Version 0.10.0 ist gerade veröffentlicht und kann nun wie folgt konfiguriert werden:
```
{
        "id" : "myInverter-2",
        "displayName" : "Photovoltaik",
        "ip" : "10.0.1.31",
        "port" : 1502,
        "updateInterval" : 30,
        "powerAddress" : 40206,
        "powerSfAddress" : 40210,
        "powerUnsignedValue" : false
}
```
Standardmäßig sind die Adressen 40083 und 40084 und unsigned ist true. Das würde den Inverter messen. Ich habe die Protokoll-Spezifikation nochmal gelesen und vermute, dass
- Meter 1 mit 40206, 40210 und false
- Meter 2 mit 40380, 40384 und false
- Meter 3 mit 40555, 40559 und false 
gelesen werden müssten.

**Comment by Morcman** [comment]
Ich hab mir das heute mal angeguckt. Die Adressen für AC Leistung (40083 und 40084) liefern tatsächlich nur die vom WR erzeugte Strommenge zum Haus. Selbst wenn man auf die Adressen für DC Leistung (40100 und 40101) wechselt, ändert das zunächst einmal nichts. Interessant ist allerdings, dass er für 40083 und 40084 Werte liefert, wenn Strom der Batterie entnommen wird. DC Werte sind dann allerdings auch gefüllt. Es muss also für den Speicher einen eigenen Adressenbereich geben. Leider finde ich dazu nichts in der Dokumentation. Ich wollte die Tage mal bei anderen Smarthome Programmen schauen, ob man dort in den PlugIn Dokus was findet.

**Comment by Kubiac** [comment]
Also es macht natürlich Sinn, dass die Werte am WR sowohl Strom vom Dach als auch aus der Batterie messen. Beide müssen ja durch den WR durch. Ich würde immer eher den AC als den DC Strom nehmen, da ersterer schon Wandlungsverluste beinhalten sollte. Hast du die anderen 3 Meter von oben mal probiert?

**Comment by eSportler89** [comment]
Moin moin,
ich hab genau die gleichen Ergebnisse wie Morcman....
40083 und 40084 scheinen für mich sogar eine kumulierte Angabe von sämtlicher durch den Wechselrichter fließender Energie zu sein, sofern diese nicht vom Versorger bezogen wird.

Meter 1 liefert bei mir keine Fehlermeldungen, jedoch auch keine Daten.
Meter 2 und 3 werden mir direkt als ungültige Registerwerte ausgegeben.


Testweise hatte ich mir mal eine VM mit Hassio aufgesetzt und dem entsprechenden Solaredge Plugin, wo ich dann gute Brauchbare Ergebnisse zurückbekommen habe. Die Dokumentation von diesem Plugin verweist auf folgende Dokumentationen von Solaredge:
[Solaredge Sunspec Technical Note](https://www.solaredge.com/sites/default/files/sunspec-implementation-technical-note.pdf)
[Power Control Protocol for Solaredge Inverters Technical Note](https://www.photovoltaikforum.com/core/attachment/88445-power-control-open-protocol-for-solaredge-inverters-pdf/)

Erstere Dokumentation ist die gleiche welche du wohl nutzt, die zweite bietet noch Optionen zu Batterie an.
Eigentlich müsste aber mit den Registern aus deiner Dokumentation auch alles andere auszulesen sein, warum Meter 1 bei mir dann nichts anzeigt ist mir noch ein Rätsel... da muss ich nochmal was ausprobieren.
Evtl. muss man mit diversen Registerwerten auch rechnen um einzelne Werte zu bekommen.

Vorstellen würde ich mir einzelne Werte zu folgenden Punkten als Lampen im HomeKit:
- Energie welche vom Dach kommt
- Energie welche das Haus verbraucht
- Energie welche zum Versorger eingespeist wird (PV Überschuss)
- Bezug vom Versorger (weil negativ LUX nicht existiert)
- Energie welche in die Batterie fließt
- Von Batterie zurück ins Haus (weil negativ LUX nicht existiert)

**Comment by Kubiac** [comment]
Ja, die SunSpec Technical Note hatte ich genutzt. Die andere kannte ich noch nicht, aber habe beim überfliegen kein Hinweis gefunden. Ich sehe mir sie mal in Ruhe an.

Das verhalten von Meter 1 müssen wir nochmal genauer untersuchen. Bei mir (ohne Batterie und Power Meter) kommt da auch schon eine Fehlermeldung. Also da ist was bei dir. Kan nes sein, dass die Entladung der Batterie dafür sorgt, dass der Meter wirklich ziemlich genau bei 0 steht? Kannst du die Werte nochmal bei Überschuss und voller Batterie und/oder Dunkelheit und leerer Batterie checken?

Das Problem mit den negativen Werten hatte ich noch gar nicht auf dem Schirm. Ja, man muss vermutlich mehrere Lichtsensoren für Export/Import und Ladung/Entladung einrichten.

Kannst du Homebridge mal im Debug-Modus starten (Drei-Punkte-Menü rechts oben - Homebridge Settings) und mir die Logs eines Ausleseversuchs schicken? Da sehe ich welche Daten er liest und was er ausrechnet.

**Comment by eSportler89** [comment]
Ich hatte die Versuche gestern bei vollem Sonnenschein gemacht. Sprich die Batterie wurde zeitweise noch geladen. Auch den Zustand Batterie voll und PV Überschuss geht ins Netz hatte ich abgedeckt.

Aktuell habe ich das Problem das ich nur noch die Meldung "[SolaredgeRealTime] connect ECONNREFUSED 192.168.1.15:1502" bekomme. Inverter habe ich schon neu gestartet, werde gleich mal in die Einstellungen gehen und schauen ob die Modbus TCP Funktion wieder ausgeschaltet ist.
Wenn ich das wieder im Griff habe, kann ich gerne mal den Debug Modus starten. Mal schauen wie viel da durch rauscht :D Meine Homebridge Instanz ist relativ groß.

**Comment by Morcman** [comment]
Über HomeAssistant habe ich auch etwas gutes dazu gefunden: https://community.home-assistant.io/t/updated-solaredge-modbus-full-setup-guide-with-energy-dashboard-integration/340956
Dort wird auch das oben genannte Plugin verwendet. Allerdings rechnet er noch einiges zusammen und auseinander, da die Werte, die vom modbus kommen wohl nicht ganz korrekt sind. Das macht das ganze Thema natürlich unendlich komplex.

**Comment by eSportler89** [comment]
Leider bekomme ich die Verbindung von der Homebridge zum Wechselrichter gar nicht mehr hin.
Modbus TCP war aktiv, auch das habe ich deaktiviert und wieder aktiviert aber keine Besserung.
Aktuell kann ich mir noch nicht erklären warum der so rumzickt.

Melde mich sobald ich das wieder in den Griff bekommen habe!

**Comment by Morcman** [comment]
Wechselrichter mal komplett neugestartet?

**Comment by eSportler89** [comment]
Jepp, das hatte ich letztens schon als Quick&Dirty Lösung probiert.

Denke das ich mal den Support anhauen muss wenn sich da nichts mehr ändert.

**Comment by Kubiac** [comment]
Ok, das hilft mir erstmal. Ich muss das mit den negativen Werten anders lösen. Ich melde mich, wenn ich ein Update habe.

**Comment by Kubiac** [comment]
Sorry fürs warten, aber andere Themen und dann war die Entwicklungsumgebung auch noch hin.

Egal Version 0.11 kann jetzt mit beliebigen Adressen konfigurieren und über den Parameter `"powerUnsignedValue" : false` auch dafür sorgen, dass negative Werte über einen zweiten Lichtsensor ausgegeben werden. Das ist für den Messpunkt am Hausanschluss und an der Batterie natürlich nötig.

Könnt ihr das bitte konfigurieren und bestä†igen, dass ihr 2 Lichtsensoren seht und dann könntet ihr die Adressen ändern und gucken, welche für welchen Messpunkt gelten.

**Comment by SchemaNature** [comment]
Wo seid Ihr mit diesem Thema gelanded? Koennt ihr bitte eine Konfiguration teilen die fuer Euch funktioniert?

**Comment by malwcal** [comment]
Hi, I can't tell if this is resolved or not: is this plugin supposed to expose consumption.  (Unfortunately I don't speak german)
